pkg/cloud-provider/cloudapi/azure/azure_security.go

```diff
@@ -83,51 +113,58 @@ func (computeCfg *computeServiceConfig) processAppliedToMembership(appliedToGrou
 		nwIntfIDLowerCase := strings.ToLower(*networkInterface.ID)
 		// 	for network interfaces not attached to any virtual machines, skip processing
-		if emptyString(networkInterface.VirtualMachineID) {
+		if networkInterface.Properties.VirtualMachine == nil || emptyString(networkInterface.Properties.VirtualMachine.ID) {
 			continue
 		}
 
 		isNsgAttached := false
-		if networkInterface.NetworkSecurityGroupID != nil && len(*networkInterface.NetworkSecurityGroupID) > 0 {
-			nsgID := strings.ToLower(*networkInterface.NetworkSecurityGroupID)
+		if networkInterface.Properties.NetworkSecurityGroup != nil && !emptyString(networkInterface.Properties.NetworkSecurityGroup.ID) {
+			nsgID := strings.ToLower(*networkInterface.Properties.NetworkSecurityGroup.ID)
 			_, _, nsgNameLowercase, err := extractFieldsFromAzureResourceID(nsgID)
 			if err != nil {
 				azurePluginLogger().Error(err, "nsg ID format not valid", "nsgID", nsgID)
 				return err
 			}
 
-			attachedAsgIDs := networkInterface.ApplicationSecurityGroupIDs
-			for _, asgID := range attachedAsgIDs {
-				asgIDLowerCase := strings.ToLower(*asgID)
-				// proceed only if network-interface attached to nephe ASG
-				_, _, asgName, err := extractFieldsFromAzureResourceID(asgIDLowerCase)
-				if err != nil {
+			for _, ipConfigs := range networkInterface.Properties.IPConfigurations {
+				if ipConfigs.Properties == nil {
 					continue
 				}
-				_, _, isAT := securitygroup.IsNepheControllerCreatedSG(asgName)
-				if isAT {
-					if strings.Compare(nsgNameLowercase, perVnetNsgNameLowercase) == 0 &&
-						strings.Compare(cloudSgNameLowercase, asgName) == 0 {
-						isNsgAttached = true
-						break
+				for _, asg := range ipConfigs.Properties.ApplicationSecurityGroups {
+					asgIDLowerCase := strings.ToLower(*asg.ID)
+					// proceed only if network-interface attached to nephe ASG
+					_, _, asgName, err := extractFieldsFromAzureResourceID(asgIDLowerCase)
+					if err != nil {
+						continue
+					}
+					_, _, isAT := securitygroup.IsNepheControllerCreatedSG(asgName)
+					if isAT {
+						if strings.Compare(nsgNameLowercase, perVnetNsgNameLowercase) == 0 &&
+							strings.Compare(cloudSgNameLowercase, asgName) == 0 {
+							isNsgAttached = true
+							break
+						}
 					}
 				}
+				// Handling just first available ip-configuration; with an assumption that it will be primary IP always.
+				break
 			}
 		}
-		_, isNicAttachedToMemberVM := memberVirtualMachines[strings.ToLower(*networkInterface.VirtualMachineID)]
+
+		_, isNicAttachedToMemberVM := memberVirtualMachines[strings.ToLower(*networkInterface.Properties.VirtualMachine.ID)]
 		_, isNicMemberNetworkInterface := memberNetworkInterfaces[strings.ToLower(*networkInterface.ID)]
 		if isNsgAttached {
 			if !isNicAttachedToMemberVM && !isNicMemberNetworkInterface {
-				nwIntfIDSetNsgToDetach[nwIntfIDLowerCase] = struct{}{}
+				nwIntfIDSetNsgToDetach[nwIntfIDLowerCase] = &networkInterface.Interface
```

Comment:
Here we are assigning the address of a local variable, the memory can get overiden when reading 2nd or 3rd interfaces.. 

This has to be changed like this.
for i := range networkInterfaces {
&networkInterfaces[i]

Reply:
not really.. &networkInterface.Interface.. networkInterface is of type 
type networkInterfaceInternal struct {
	armnetwork.Interface
	vnetID string
} and address of armnetwork.Interface is passed

Reply:
If it was just &networkInterface, then that could have caused issue, but here it is address on struct member, so it would be fine.